File: new_practice/199.binary-tree-right-side-view.py
#
# @lc app=leetcode id=199 lang=python3
#
# [199] Binary Tree Right Side View
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def rightSideView(self, root: TreeNode) -> List[int]:
    def rightSideView(self, root):

        if not root:
            return []

        
        level = [root]
        ans = []

        while level:
            ans.append(level[-1].val)

            next_level = []
            for ele in level:
                if ele.left is not None:
                    next_level.append(ele.left)
                if ele.right is not None:
                    next_level.append(ele.right)

            level = next_level

        return ans








        
# @lc code=end

File: new_practice_cpp/199.binary-tree-right-side-view.cpp
/*
 * @lc app=leetcode id=199 lang=cpp
 *
 * [199] Binary Tree Right Side View
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


// level order
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if(!root)
            return {};
            
        vector<int> ans;
        vector<TreeNode*> level_nodes;
        level_nodes.push_back(root);

        while(!level_nodes.empty()) {
            ans.push_back(level_nodes[level_nodes.size()-1]->val); // push ans

            vector<TreeNode*> next_level;
            for(int i=0; i<level_nodes.size();i++){
                if(level_nodes[i]->left!=NULL)
                    next_level.push_back(level_nodes[i]->left);
                if(level_nodes[i]->right!=NULL)
                    next_level.push_back(level_nodes[i]->right);
            }

            level_nodes = next_level;
        }
        return ans;
    }
};
// @lc code=end

